Add Django to CI and add staging infrastructure

BryanQuigley opened the pull request:
Environment name stgdjango
Externally stg.echosearch.org

Using aws 4.8 and terraform 1.0
Move from azavea cert, ecr and rds
New cibuild/cipublish that works with Django images
Latest images will be used by default when ever plan/apply is run per force_new_deployment
Dropped travisCI
Add Github CI

Items to look at in future:
deployment/terraform-django/app.tf
 * healthcheck is healthy when 400 response (change to 200 eventually)
 * hardcoded to staging as our environment name is stg.
Dropped automatic cert validation terraform
	due to inability to provision staging.es worked around with name change
	staging -> stg.  staging failed with aws console too

Resolves #415

## Testing Instructions

 * Look out CI output to confirm it's doing the correct thing.

BryanQuigley committed:
Add Django to CI and add staging infrastructure

Environment name stgdjango
Externally stg.echosearch.org

Using aws 4.8 and terraform 1.0
Move away from azavea cert, ecr and rds
New cibuild/cipublish that works with Django images
Latest images will be used by default when ever plan/apply is run per force_new_deployment
Dropped travisCI
Add Github CI
 * do staging deploys on develop or test/**
 * do cibuild on any other PR

Items to look at in future:
deployment/terraform-django/app.tf
 * healthcheck is healthy when 400 response (change to 200 eventually)
 * hardcoded to staging as our environment name is stg.
Dropped automatic cert validation terraform
	due to inability to provision staging.es worked around with name change
	staging -> stg.  staging failed with aws console too
Althogh provisioned I have not extensiely tested:
 * Bastion (was hoping to kill it)
 * AppCLI

Resolves #415

ddohler commented:
I discussed with @maurizi and he's going to take a look at this while I'm out, but may leave full review until I return.

ddohler commented:
Also adding @ktohalloran for secondary review since @aaronxsu is out next week too.

maurizi commented on deployment/README-django.md:
```diff
@@ -0,0 +1,89 @@
+- [AWS Credentials](#aws-credentials)
+- [Publish Container Images](#publish-container-images)
+- [Terraform](#terraform)
+- [Bastion](#bastion)
+  - [PostgreSQL Client](#postgresql-client)
```
These links are not working - did you intend to write these sections and forget, or should the links just be removed from the ToC?

BryanQuigley replied:
removed but eventually want to add a bastion-like section

maurizi commented on deployment/terraform-django/app.tf:
```diff
@@ -0,0 +1,219 @@
+  health_check {
+    healthy_threshold   = "3"
+    interval            = "30"
+    matcher             = "400"
```
I presume this is simply because we don't have a `/health-check/` endpoint at all yet, so the application is returning 404?

Can you create an issue for adding healthchecks (it looks like on other Django projects we typically use https://github.com/mwarkentin/django-watchman) and add a `TODO` here that references it?

BryanQuigley replied:
Done, https://github.com/azavea/echo-locator/issues/441

maurizi replied:
Please add the TODO pointing to the Github issue as well
This is a team process we try to follow where every TODO refs a Github issue, so that
1) the TODOs are tracked by issues for planning purposes so we don't forget about them
2) you can easily search for all the TODOs that reference an issue when you're working on it

```suggestion
    # TODO GH #411: Set this to 200 after we add a health check
    matcher             = "400"
```

BryanQuigley replied:
oh, did a different format, will redo with that^

maurizi commented on deployment/terraform-django/cdn.tf:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  enabled         = true
+  is_ipv6_enabled = true
```
🎉

maurizi commented on deployment/terraform-django/dashboards/main.json.tmpl:
```diff
@@ -0,0 +1,326 @@
+              "color": "#dbdb8d"
+            }
+          ],
+          [
+            ".",
+            "HTTPCode_Target_4XX_Count",
+            ".",
+            ".",
+            {
+              "color": "#ff7f0e"
```
Will this alarm be an issue with the health check returning 4xx status codes? I guess it's an incentive to get the health check implemented...

BryanQuigley replied:
here, it's actually just a metric that we aren't actually alarming on..

BryanQuigley commented:
stg.echosearch.org NS record from echosearch.org needed to be created manually. I'm relatively sure this is what broke automatic cert validation (but that will only be an issue IMHO if we change the domain again).

ktohalloran commented on deployment/README-django.md:
````diff
@@ -0,0 +1,85 @@
+
+```bash
+$ aws configure --profile echo-locator
+AWS Access Key ID [None]: AKIAIOSFODNN7EXAMPLE
+AWS Secret Access Key [None]: wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
````
I found this to be a bit confusing--with all the gibberish letters I missed the "EXAMPLE"  and "EXAMPLEKEY" at the end; I think it might be more self-explanatory to use
```
AWS Access Key ID [None]: <your AWS Key>
AWS Secret Access Key [None]: <your AWS Secret Access Key>
```

On a related note, what AWS credentials should these be, e.g. do we have project-wide AWS credentials or will you need individual IAM credentials assigned by the project lead?

BryanQuigley replied:
Will change to above for now.

ktohalloran commented on deployment/README-django.md:
```diff
@@ -0,0 +1,85 @@
+...
+Successfully built cc2b35ef78c4
+Successfully tagged echolocator:latest
+export AWS_PROFILE=echo-locator
```
Is this `export AWS_PROFILE=echo-locator` a command that needs to be run in between `./scripts/cibuild` and `./scripts/cipublish` or is it just a statement echoed to the console?

BryanQuigley replied:
you have to run that, so adding $, thanks!

ktohalloran commented on deployment/terraform-django/app.tf:
```diff
@@ -0,0 +1,219 @@
+  }
+}
+
+locals {
+  django_container_name = "django"
+  django_container_port = 8085
+}
```
Could these be consolidated with the `locals` section of `variables.tf`?

BryanQuigley replied:
Good point - they actually likely should be moved out of that directory entirely, opened a new issue for it - https://github.com/azavea/echo-locator/issues/451
(I haven't seen any of our projects do that though)

ktohalloran commented on deployment/terraform-django/storage.tf:
```diff
@@ -0,0 +1,98 @@
+locals {
+  site_bucket_name = lower("${var.project}-${var.environment}-tiles-${var.aws_region}")
+  logs_bucket_name = lower("${var.project}-${var.environment}-logs-${var.aws_region}")
+}
```
Same as above, could these be consolidated into `locals` inside `variables.tf`?

BryanQuigley replied:
this is only used in that tf file so it makes sense to stay there.

maurizi commented on deployment/terraform-django/database.tf:
```diff
@@ -0,0 +1,163 @@
+# Security group resources
+#
+resource "aws_security_group" "postgresql" {
+vpc_id = module.vpc.id
```
```suggestion
  vpc_id = module.vpc.id
```

maurizi commented on deployment/terraform-django/storage.tf:
```diff
@@ -0,0 +1,98 @@
+
+  #data "aws_cloudfront_log_delivery_canonical_user_id" "cdn" {}
+  # TODO I guess we need to create this grant when we have an ID manually?
+  # TODO Confirm works per https://github.com/terraform-aws-modules/terraform-aws-s3-bucket/pull/113/files
```
Can you create a follow-on issue for this, and reference the issue # in the TODO?

maurizi commented on scripts/cipublish:
```diff
@@ -10,38 +10,51 @@ function usage() {
+function amazon_ecr_login() {
+    # Retrieves a temporary authorization token that can be used to access
+    # Amazon ECR, along with the registry URL.
+    read -r AUTHORIZATION_TOKEN ECR_REGISTRY \
+        <<<"$(aws ecr get-authorization-token \
+            --output "text" \
+            --query "authorizationData[0].[authorizationToken, proxyEndpoint]")"
```
👀 I'm going to have to reread the bash scripting guide, this is doing some fancy stuff I haven't seen before, I like it!

BryanQuigley replied:
I can't take credit for this, guessing thanks go to other ops members

ktohalloran commented on scripts/cipublish:
```diff
@@ -10,38 +10,51 @@ function usage() {
-        # Move the Amplify config for the proper env into the source dir
-        cp "deployment/amplify/${ENV}/src/aws-exports.js" taui/src/aws-exports.js
+function amazon_ecr_login() {
+    # Retrieves a temporary authorization token that can be used to access
+    # Amazon ECR, along with the registry URL.
```
These comments are so helpful! Thanks for writing these in, it definitely helps this file's readability.

BryanQuigley replied:
Props to former ops team members!

maurizi reviewed (approved):
This is looking really good! I left some comments for various cleanups but they're all fairly trivial - feel free to merge this once they're taken care of.

ktohalloran reviewed (commented):
This looks great and I'm really impressed at the size at the task, that was a lot of infrastructure to build/replace at once! I had a few questions, particularly around how Taui fits into this picture.  I was surprised to see the `taui` service removed from `docker-compose.ci.yml`, so can you please explain a bit more about how this infrastructure deploys the frontend? I'm not sure if [issue 376](https://github.com/azavea/echo-locator/issues/376) will impact this infrastructure much at all or if this transition is going to be an ongoing process, but I think if this is the permanent method of deployment, it also might be worth adding a note about this in the deployment `README`.

Also, how do the `terraform` and `terraform-django` folders relate to each other? Is that initial `terraform` folder still necessary? If so, I think adding some kind of note to either of the `README`s explaining that relationship might be helpful. If not, is it within scope of this issue to remove that old `terraform` folder? 

Finally, I didn't see `./scripts/test` referenced anywhere within `.github/workflows/ci.yml`. Are you using another method of checking for linting/formatting errors in both Taui and Django while building?

BryanQuigley committed:
Updated for review comments

 - add TODO for adding health check (#441)
 - drop bastion broken link and added issue to finish bastion like access (#449)
 - added issue to remove the old deployment bits (#448)
 - terraform formatting
 - d/Readme-django improvments (also note in d/Readme that that setup is going away)
 - still needs ./scripts/test added including django/lint (#452)
 - add issue for cloudfront logging (#453)

BryanQuigley commented:
> infrastructure deploys the frontend?

I dropped deploying the frontend and sent out to #424.

>terraform and terraform-django

opened new issue for that in #448.  We want to keep the old terraform folder until we purge the old infra.

>Finally, I didn't see ./scripts/test referenced anywhere within .github/workflows/ci.yml

Good catch!  I originally had cibuild running the tests, but ran into issues with that and dropped it (and forgot to mention that or make a new issue). I've added back the existing ./scripts/test and will make a new issue for adding django testing.

BryanQuigley commented:
Dropped test again - it still depends on some amplify bits.

ktohalloran reviewed (approved):
Looks great, thanks for taking care of all of that!
